File: app/src/main/java/net/ktnx/mobileledger/model/HledgerVersion.java
/*
 * Copyright © 2020, 2024 Damyan Ivanov.
 * This file is part of MoLe.
 * MoLe is free software: you can distribute it and/or modify it
 * under the term of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your opinion), any later version.
 *
 * MoLe is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License terms for details.
 *
 * You should have received a copy of the GNU General Public License
 * along with MoLe. If not, see <https://www.gnu.org/licenses/>.
 */

package net.ktnx.mobileledger.model;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import net.ktnx.mobileledger.json.API;

import java.util.Locale;

public class HledgerVersion {
    private final int major;
    private final int minor;
    private final int patch;
    private final boolean isPre_1_20_1;
    private final boolean hasPatch;
    public HledgerVersion(int major, int minor) {
        this.major = major;
        this.minor = minor;
        this.patch = 0;
        this.isPre_1_20_1 = false;
        this.hasPatch = false;
    }
    public HledgerVersion(int major, int minor, int patch) {
        this.major = major;
        this.minor = minor;
        this.patch = patch;
        this.isPre_1_20_1 = false;
        this.hasPatch = true;
    }
    public HledgerVersion(boolean pre_1_20_1) {
        if (!pre_1_20_1)
            throw new IllegalArgumentException("pre_1_20_1 argument must be true");
        this.major = this.minor = this.patch = 0;
        this.isPre_1_20_1 = true;
        this.hasPatch = false;
    }
    public HledgerVersion(HledgerVersion origin) {
        this.major = origin.major;
        this.minor = origin.minor;
        this.isPre_1_20_1 = origin.isPre_1_20_1;
        this.patch = origin.patch;
        this.hasPatch = origin.hasPatch;
    }
    @Override
    public boolean equals(@Nullable Object obj) {
        if (obj == null)
            return false;
        if (!(obj instanceof HledgerVersion))
            return false;
        HledgerVersion that = (HledgerVersion) obj;

        return (this.isPre_1_20_1 == that.isPre_1_20_1 && this.major == that.major &&
                this.minor == that.minor && this.patch == that.patch &&
                this.hasPatch == that.hasPatch);
    }
    public boolean isPre_1_20_1() {
        return isPre_1_20_1;
    }
    public int getMajor() {
        return major;
    }
    public int getMinor() {
        return minor;
    }
    public int getPatch() {
        return patch;
    }
    @NonNull
    @Override
    public String toString() {
        if (isPre_1_20_1)
            return "(before 1.20)";
        return hasPatch ? String.format(Locale.ROOT, "%d.%d.%d", major, minor, patch)
                        : String.format(Locale.ROOT, "%d.%d", major, minor);
    }
    public boolean atLeast(int major, int minor) {
        return ((this.major == major) && (this.minor >= minor)) || (this.major > major);
    }
    @org.jetbrains.annotations.Nullable
    public API getSuitableApiVersion() {
        if (isPre_1_20_1)
            return null;

        // Return the most appropriate API version based on detected hledger-web version
        // Versions are checked in descending order to select the highest compatible API
        if (atLeast(1, 50)) {
            return API.v1_50;
        }
        else if (atLeast(1, 40)) {
            return API.v1_40;
        }
        else if (atLeast(1, 32)) {
            return API.v1_32;
        }
        else if (atLeast(1, 23)) {
            return API.v1_23;
        }
        else if (atLeast(1, 19)) {
            return API.v1_19_1;
        }
        else if (atLeast(1, 15)) {
            return API.v1_15;
        }
        else if (atLeast(1, 14)) {
            return API.v1_14;
        }

        // For versions older than 1.14, return null (not supported)
        return null;
    }
}

File: app/src/main/kotlin/net/ktnx/mobileledger/model/HledgerVersion.kt
/*
 * Copyright © 2020, 2024 Damyan Ivanov.
 * This file is part of MoLe.
 * MoLe is free software: you can distribute it and/or modify it
 * under the term of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your opinion), any later version.
 *
 * MoLe is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License terms for details.
 *
 * You should have received a copy of the GNU General Public License
 * along with MoLe. If not, see <https://www.gnu.org/licenses/>.
 */

package net.ktnx.mobileledger.model

import java.util.Locale
import net.ktnx.mobileledger.json.API

class HledgerVersion {
    val major: Int
    val minor: Int
    val patch: Int
    val isPre_1_20_1: Boolean
    private val hasPatch: Boolean

    constructor(major: Int, minor: Int) {
        this.major = major
        this.minor = minor
        this.patch = 0
        this.isPre_1_20_1 = false
        this.hasPatch = false
    }

    constructor(major: Int, minor: Int, patch: Int) {
        this.major = major
        this.minor = minor
        this.patch = patch
        this.isPre_1_20_1 = false
        this.hasPatch = true
    }

    constructor(pre_1_20_1: Boolean) {
        require(pre_1_20_1) { "pre_1_20_1 argument must be true" }
        this.major = 0
        this.minor = 0
        this.patch = 0
        this.isPre_1_20_1 = true
        this.hasPatch = false
    }

    constructor(origin: HledgerVersion) {
        this.major = origin.major
        this.minor = origin.minor
        this.isPre_1_20_1 = origin.isPre_1_20_1
        this.patch = origin.patch
        this.hasPatch = origin.hasPatch
    }

    override fun equals(other: Any?): Boolean {
        if (other == null) return false
        if (other !is HledgerVersion) return false

        return isPre_1_20_1 == other.isPre_1_20_1 &&
                major == other.major &&
                minor == other.minor &&
                patch == other.patch &&
                hasPatch == other.hasPatch
    }

    override fun hashCode(): Int {
        var result = major
        result = 31 * result + minor
        result = 31 * result + patch
        result = 31 * result + isPre_1_20_1.hashCode()
        result = 31 * result + hasPatch.hashCode()
        return result
    }

    override fun toString(): String = if (isPre_1_20_1) {
            "(before 1.20)"
        } else if (hasPatch) {
            String.format(Locale.ROOT, "%d.%d.%d", major, minor, patch)
        } else {
            String.format(Locale.ROOT, "%d.%d", major, minor)
        }

    fun atLeast(major: Int, minor: Int): Boolean = (this.major == major && this.minor >= minor) || this.major > major

    fun getSuitableApiVersion(): API? {
        if (isPre_1_20_1) return null

        return when {
            atLeast(1, 50) -> API.v1_50
            atLeast(1, 40) -> API.v1_40
            atLeast(1, 32) -> API.v1_32
            atLeast(1, 23) -> API.v1_23
            atLeast(1, 19) -> API.v1_19_1
            atLeast(1, 15) -> API.v1_15
            atLeast(1, 14) -> API.v1_14
            else -> null
        }
    }
}
